torchaudio/datasets/musdb_hq.py

```diff
@@ -0,0 +1,93 @@
+import os
+from pathlib import Path
+from typing import Tuple, Union, List, Optional
+
+import torch
+import torchaudio
+from torch import Tensor
+from torch.utils.data import Dataset
+from torchaudio import transforms as T
+
+URL = "https://zenodo.org/record/3338373/files/musdb18hq.zip"
+EXT = ".wav"
+
+
+class MUSDB_HQ(Dataset):
+    def __init__(
+        self,
+        root: Union[str, Path],
+        sources: List[str],
+        subset: str,
+        samplerate: int = 44100,
+        split: Optional[str] = None,
+        validation: Optional[List[str]] = None,
+    ) -> None:
+        self.sources = sources
+        self.user_samplerate = samplerate
+        self.split = split
+        self.validation = validation
+
+        basename = os.path.basename(URL)
+        basename = basename.rsplit(".", 2)[0]
+        assert subset in ["test", "train"], "`subset` must be one of ['test', 'train']"
+        assert self.split is None or self.split in [
+            "train",
+            "validation",
+        ], "`split` must be one of ['train', 'validation']"
+        assert (
+            self.split is None and self.validation is None or self.split and self.validation
+        ), "`split` and `validation` must both be None or not be None"
+        self._path = os.path.join(root, basename)
+        self._path = os.path.join(self._path, subset)
+        self.names = self.metadata_build()
+
+    def get_file(self, name, source):
+        return Path(self._path) / name / f"{source}{EXT}"
+
+    def _load_sample(self, n: int) -> Tuple[torch.Tensor, int, str]:
+        name = self.names[n]
+        wavs = []
+        sample_rate = None
+        for source in self.sources:
+            file = self.get_file(name, source)
+            wav, new_samplerate = torchaudio.load(str(file))
+            if sample_rate is None:
+                sample_rate = new_samplerate
+            elif sample_rate != new_samplerate:
+                raise ValueError(
+                    f"Invalid sample rate for file {file}: " f"expecting {sample_rate} but got {new_samplerate}."
+                )
+            wavs.append(wav)
+
+        stacked = torch.stack(wavs)
+        transform = T.Resample(sample_rate, self.user_samplerate)
```

Comment:
This constructs resampling kernel every time a sample is fetched from dataset.
Either construct the transform in constructor and cache, or let client code handle the resampling.